batchlog manager is not stopped if scylla shuts down during startup

This issue stems off of #8995.

In addition to storage_proxy::drain_on_shutdown not synchronizing with on-going queries, the reason there is an on-going query after the storage_proxy is drained appears to be that the batchlog manager is not stopped if scylla gets a shutdown signal during startup (e.g. when waiting for gossip to settle).

I'm opening a separate issue for that since I think we should fix the storage_proxy draining synchronization as well, but this issue is orthogonal and fixing it will also fix the failures we see in several dtests, like the following one:

https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/864/testReport/update_cluster_layout_tests/TestUpdateClusterLayout/simple_add_new_node_while_adding_info_1_test/
```
Traceback (most recent call last):
  File "/usr/lib64/python3.7/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib64/python3.7/unittest/case.py", line 648, in run
    self.tearDown()
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/dtest.py", line 1148, in tearDown
    self.cleanUpCluster()
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/dtest.py", line 1202, in cleanUpCluster
    raise AssertionError('Unexpected errors found: {}'.format(found_errors))
AssertionError: Unexpected errors found: [('node4', ['ERROR 2021-07-05 02:43:39,078 [shard 0] storage_proxy - Exception when communicating with 127.0.53.4, to read from system.batchlog: seastar::gate_closed_exception (gate closed)', 'ERROR 2021-07-05 02:43:39,079 [shard 0] batchlog_manager - Exception in batch replay: exceptions::read_failure_exception (Operation failed for system.batchlog - received 0 responses and 1 failures from 1 CL=ONE.)'])]
```

https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/864/artifact/logs-all.release.2/1625453021390_update_cluster_layout_tests.TestUpdateClusterLayout.simple_add_new_node_while_adding_info_1_test/node4.log
```
INFO  2021-07-05 02:43:33,373 [shard 0] init - Shutting down drain storage proxy
INFO  2021-07-05 02:43:33,373 [shard 0] hints_manager - Asked to stop
INFO  2021-07-05 02:43:33,373 [shard 0] hints_manager - Stopped
INFO  2021-07-05 02:43:33,374 [shard 1] hints_manager - Asked to stop
INFO  2021-07-05 02:43:33,374 [shard 1] hints_manager - Stopped
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down drain storage proxy was successful
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down streaming service
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down streaming service was successful
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down Raft RPC
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down Raft RPC was successful
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down storage proxy RPC verbs
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down storage proxy RPC verbs was successful
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down sstables format selector
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down sstables format selector was successful
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down migration manager
INFO  2021-07-05 02:43:33,374 [shard 0] migration_manager - stopping migration service
INFO  2021-07-05 02:43:33,374 [shard 1] migration_manager - stopping migration service
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down migration manager was successful
INFO  2021-07-05 02:43:33,374 [shard 0] init - Shutting down database
INFO  2021-07-05 02:43:39,078 [shard 0] init - Shutting down database was successful
INFO  2021-07-05 02:43:39,078 [shard 0] init - Shutting down service_memory_limiter
INFO  2021-07-05 02:43:39,078 [shard 0] init - Shutting down service_memory_limiter was successful
INFO  2021-07-05 02:43:39,078 [shard 0] init - Shutting down sst_dir_semaphore
ERROR 2021-07-05 02:43:39,078 [shard 0] storage_proxy - Exception when communicating with 127.0.53.4, to read from system.batchlog: seastar::gate_closed_exception (gate closed)
ERROR 2021-07-05 02:43:39,079 [shard 0] batchlog_manager - Exception in batch replay: exceptions::read_failure_exception (Operation failed for system.batchlog - received 0 responses and 1 failures from 1 CL=ONE.)
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down sst_dir_semaphore was successful
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down Raft
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down Raft was successful
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down messaging service
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down messaging service was successful
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down API server
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down API server was successful
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down tracing instance
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down tracing instance was successful
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down migration manager notifier
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down migration manager notifier was successful
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down prometheus API server
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down prometheus API server was successful
INFO  2021-07-05 02:43:39,079 [shard 0] init - Shutting down sighup
INFO  2021-07-05 02:43:39,080 [shard 0] init - Shutting down sighup was successful
ERROR 2021-07-05 02:43:39,080 [shard 0] init - Startup failed: seastar::sleep_aborted (Sleep is aborted)
```

--- a/db/batchlog_manager.cc
+++ b/db/batchlog_manager.cc
@@ -152,6 +152,10 @@ future<> db::batchlog_manager::stop() {
     if (!_stop.abort_requested()) {
         _stop.request_abort();
     }
+    if (this_shard_id() == 0) {
+        // Abort do_batch_log_replay if waiting on the semaphore.
+        _sem.broken();
+    }
     return std::exchange(_started, make_ready_future<>()).finally([this] {
         return _gate.close();
     });

--- a/db/batchlog_manager.cc
+++ b/db/batchlog_manager.cc
@@ -43,6 +43,8 @@
 #include <seastar/core/do_with.hh>
 #include <seastar/core/semaphore.hh>
 #include <seastar/core/metrics.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sleep.hh>
 #include <boost/range/adaptor/map.hpp>
 #include <boost/range/adaptor/sliced.hpp>
 
@@ -79,6 +81,7 @@ db::batchlog_manager::batchlog_manager(cql3::query_processor& qp, batchlog_manag
         : _qp(qp)
         , _write_request_timeout(std::chrono::duration_cast<db_clock::duration>(config.write_request_timeout))
         , _replay_rate(config.replay_rate)
+        , _started(make_ready_future<>())
         , _delay(config.delay) {
     namespace sm = seastar::metrics;
 
@@ -114,35 +117,44 @@ future<> db::batchlog_manager::do_batch_log_replay() {
     });
 }
 
+future<> db::batchlog_manager::batchlog_replay_loop() {
+    assert (this_shard_id() == 0);
+    auto delay = _delay;
+    while (!_stop.abort_requested()) {
+        try {
+            co_await sleep_abortable(delay, _stop);
+        } catch (sleep_aborted&) {
+            co_return;
+        }
+        try {
+            co_await do_batch_log_replay();
+        } catch (...) {
+            blogger.error("Exception in batch replay: {}", std::current_exception());
+        }
+        delay = std::chrono::milliseconds(replay_interval);
+    }
+}
+
 future<> db::batchlog_manager::start() {
     // Since replay is a "node global" operation, we should not attempt to do
     // it in parallel on each shard. It will just overlap/interfere.  To
-    // simplify syncing between the timer and user initiated replay operations,
-    // we use the _timer and _sem on shard zero only. Replaying batchlog can
+    // simplify syncing between batchlog_replay_loop and user initiated replay operations,
+    // we use the _sem on shard zero only. Replaying batchlog can
     // generate a lot of work, so we distrute the real work on all cpus with
     // round-robin scheduling.
     if (this_shard_id() == 0) {
-        _timer.set_callback([this] {
-            // Do it in the background.
-            (void)do_batch_log_replay().handle_exception([] (auto ep) {
-                blogger.error("Exception in batch replay: {}", ep);
-            }).finally([this] {
-                _timer.arm(lowres_clock::now() + std::chrono::milliseconds(replay_interval));
-            });
-        });
-
-        _timer.arm(lowres_clock::now() + _delay);
+        _started = batchlog_replay_loop();
     }
     return make_ready_future<>();
 }
 
 future<> db::batchlog_manager::stop() {
-    if (_stop) {
-        return make_ready_future<>();
+    if (!_stop.abort_requested()) {
+        _stop.request_abort();
     }
-    _stop = true;
-    _timer.cancel();
-    return _gate.close();
+    return std::exchange(_started, make_ready_future<>()).finally([this] {
+        return _gate.close();
+    });
 }
 
 future<size_t> db::batchlog_manager::count_all_batches() const {

--- a/db/batchlog_manager.cc
+++ b/db/batchlog_manager.cc
@@ -93,19 +93,23 @@ future<> db::batchlog_manager::do_batch_log_replay() {
     // Use with_semaphore is much simpler, but nested invoke_on can
     // cause deadlock.
     return get_batchlog_manager().invoke_on(0, [] (auto& bm) {
+        bm._gate.enter();
         return bm._sem.wait().then([&bm] {
             return bm._cpu++ % smp::count;
         });
     }).then([] (auto dest) {
         blogger.debug("Batchlog replay on shard {}: starts", dest);
         return get_batchlog_manager().invoke_on(dest, [] (auto& bm) {
-            return bm.replay_all_failed_batches();
+            return with_gate(bm._gate, [&bm] {
+                return bm.replay_all_failed_batches();
+            });
         }).then([dest] {
             blogger.debug("Batchlog replay on shard {}: done", dest);
         });
     }).finally([] {
         return get_batchlog_manager().invoke_on(0, [] (auto& bm) {
-            return bm._sem.signal();
+            bm._sem.signal();
+            bm._gate.leave();
         });
     });
 }

--- a/db/batchlog_manager.cc
+++ b/db/batchlog_manager.cc
@@ -148,7 +148,7 @@ future<> db::batchlog_manager::start() {
     return make_ready_future<>();
 }
 
-future<> db::batchlog_manager::stop() {
+future<> db::batchlog_manager::drain() {
     if (!_stop.abort_requested()) {
         _stop.request_abort();
     }
@@ -156,7 +156,13 @@ future<> db::batchlog_manager::stop() {
         // Abort do_batch_log_replay if waiting on the semaphore.
         _sem.broken();
     }
-    return std::exchange(_started, make_ready_future<>()).finally([this] {
+    return with_gate(_gate, [this] {
+        return std::exchange(_started, make_ready_future<>());
+    });
+}
+
+future<> db::batchlog_manager::stop() {
+    return drain().finally([this] {
         return _gate.close();
     });
 }

--- a/db/batchlog_manager.hh
+++ b/db/batchlog_manager.hh
@@ -107,6 +107,8 @@ public:
     batchlog_manager(cql3::query_processor&, batchlog_manager_config config);
 
     future<> start();
+    // abort the replay loop and return its future.
+    future<> drain();
     future<> stop();
 
     future<> do_batch_log_replay();

--- a/db/batchlog_manager.hh
+++ b/db/batchlog_manager.hh
@@ -47,6 +47,7 @@
 #include <seastar/core/timer.hh>
 #include <seastar/core/gate.hh>
 #include <seastar/core/metrics_registration.hh>
+#include <seastar/core/abort_source.hh>
 
 #include "gms/inet_address.hh"
 #include "inet_address_vectors.hh"
@@ -89,12 +90,12 @@ private:
     cql3::query_processor& _qp;
     db_clock::duration _write_request_timeout;
     uint64_t _replay_rate;
-    timer<clock_type> _timer;
+    future<> _started;
     std::chrono::milliseconds _delay;
     semaphore _sem{1};
     seastar::gate _gate;
     unsigned _cpu = 0;
-    bool _stop = false;
+    seastar::abort_source _stop;
 
     std::default_random_engine _e1{std::random_device{}()};
 
@@ -119,6 +120,8 @@ public:
     db_clock::duration get_batch_log_timeout() const;
 
     inet_address_vector_replica_set endpoint_filter(const sstring&, const std::unordered_map<sstring, std::unordered_set<gms::inet_address>>&);
+private:
+    future<> batchlog_replay_loop();
 };
 
 extern distributed<batchlog_manager> _the_batchlog_manager;

--- a/main.cc
+++ b/main.cc
@@ -1002,7 +1002,6 @@ int main(int ac, char** av) {
 
             db::get_batchlog_manager().start(std::ref(qp), bm_cfg).get();
             // #293 - do not stop anything
-            // engine().at_exit([] { return db::get_batchlog_manager().stop(); });
             sstables::init_metrics().get();
 
             db::system_keyspace::minimal_setup(qp);
@@ -1281,6 +1280,9 @@ int main(int ac, char** av) {
             db::get_batchlog_manager().invoke_on_all([] (db::batchlog_manager& b) {
                 return b.start();
             }).get();
+            auto stop_batchlog_manager = defer_verbose_shutdown("batchlog manager", [] {
+                db::get_batchlog_manager().invoke_on_all(&db::batchlog_manager::stop).get();
+            });
 
             supervisor::notify("starting load meter");
             load_meter.init(db, gms::get_local_gossiper()).get();

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -2026,7 +2026,7 @@ future<> storage_service::decommission() {
             slogger.info("DECOMMISSIONING: stopped transport");
 
             db::get_batchlog_manager().invoke_on_all([] (auto& bm) {
-                return bm.stop();
+                return bm.drain();
             }).get();
             slogger.info("DECOMMISSIONING: stop batchlog_manager done");
 
@@ -2719,7 +2719,7 @@ future<> storage_service::do_drain(bool on_shutdown) {
         flush_column_families();
 
         db::get_batchlog_manager().invoke_on_all([] (auto& bm) {
-            return bm.stop();
+            return bm.drain();
         }).get();
 
         set_mode(mode::DRAINING, "shutting down migration manager", false);

